src/main/java/org/mvplugins/multiverse/portals/MVPortal.java

```diff
@@ -63,25 +68,40 @@ public static boolean isPortalInterior(Material material) {
         return INTERIOR_MATERIALS.contains(material);
     }
 
-    public MVPortal(MultiversePortals instance, String name) {
-        init(instance, name, true);
+    public MVPortal(@NotNull MultiversePortals plugin, @NotNull String name) {
+        this(plugin, name, true);
+    }
+
+    public MVPortal(LoadedMultiverseWorld world, MultiversePortals instance, String name, String owner, String location) {
+        this(instance, name);
+        this.setOwner(owner);
+        this.setPortalLocation(location, world);
     }
 
-    private MVPortal(MultiversePortals instance, String name, boolean allowSave) {
-        init(instance, name, allowSave);
+    public MVPortal(MultiversePortals plugin, String name, String owner, PortalLocation location) {
+        this(plugin, name);
+        this.setOwner(owner);
+        this.setPortalLocation(location);
     }
 
     // If this is called with allowSave=false, the caller needs to be sure to
     // call allowSave() when they're finished modifying it.
-    private void init(MultiversePortals instance, String name, boolean allowSave) {
+    private MVPortal(
+            MultiversePortals plugin,
+            String name,
+            boolean allowSave) {
         // Disallow saving until initialization is finished.
         this.allowSave = false;
 
-        this.plugin = instance;
+        this.plugin = plugin;
+        this.worldManager = this.plugin.getServiceLocator().getService(WorldManager.class);
+        this.destinationsProvider = this.plugin.getServiceLocator().getService(DestinationsProvider.class);
+        this.blockSafety = this.plugin.getServiceLocator().getService(BlockSafety.class);
```

Comment:
Shouldn't these just be injected?

Reply:
I can’t do it bcu MVPortal can’t be a service, at least not with the current implementation. Bcu you need a instance of MVPortal for every portal created.